File: frontend/src/utils/validators.js
/**
 * Frontend Email Validation Utilities for LoanAssist.
 * Enforces RFC 5322 compliance, strict Top-Level Domain (TLD) validation,
 * internal whitespace rejection, and strict domain rules (such as @gmail.com).
 */

export const validateEmail = (rawEmail) => {
  if (!rawEmail || typeof rawEmail !== 'string') {
    return {
      isValid: false,
      error: 'Please enter a valid email address.'
    };
  }

  const trimmed = rawEmail.trim();

  if (!trimmed) {
    return {
      isValid: false,
      error: 'Please enter a valid email address.'
    };
  }

  // Reject internal whitespace (e.g., 'user @gmail.com')
  if (/\s/.test(trimmed)) {
    return {
      isValid: false,
      error: 'Please enter a valid email address.'
    };
  }

  // Must contain exactly one '@'
  const atCount = (trimmed.match(/@/g) || []).length;
  if (atCount !== 1) {
    return {
      isValid: false,
      error: 'Please enter a valid email address.'
    };
  }

  const [localPart, domain] = trimmed.split('@');

  // Both local-part and domain must be non-empty
  if (!localPart || !domain) {
    return {
      isValid: false,
      error: 'Please enter a valid email address.'
    };
  }

  // Local-part checks
  if (localPart.length > 64) {
    return {
      isValid: false,
      error: 'Please enter a valid email address.'
    };
  }

  if (
    localPart.startsWith('.') ||
    localPart.endsWith('.') ||
    localPart.includes('..')
  ) {
    return {
      isValid: false,
      error: 'Please enter a valid email address.'
    };
  }

  const localRegex = /^[a-zA-Z0-9.!#$%&'*+/=?^_`{|}~-]+$/;
  if (!localRegex.test(localPart)) {
    return {
      isValid: false,
      error: 'Please enter a valid email address.'
    };
  }

  // Domain checks
  if (domain.length > 255) {
    return {
      isValid: false,
      error: 'Please enter a valid email address.'
    };
  }

  if (
    domain.startsWith('.') ||
    domain.endsWith('.') ||
    domain.startsWith('-') ||
    domain.endsWith('-') ||
    domain.includes('..') ||
    !domain.includes('.')
  ) {
    return {
      isValid: false,
      error: 'Please enter a valid email address.'
    };
  }

  const domainLabels = domain.split('.');
  for (const label of domainLabels) {
    if (!label || label.startsWith('-') || label.endsWith('-')) {
      return {
        isValid: false,
        error: 'Please enter a valid email address.'
      };
    }
    if (!/^[a-zA-Z0-9-]+$/.test(label)) {
      return {
        isValid: false,
        error: 'Please enter a valid email address.'
      };
    }
  }

  const tld = domainLabels[domainLabels.length - 1];
  // TLD must consist only of letters and be between 2 and 24 characters
  if (!/^[a-zA-Z]{2,24}$/.test(tld)) {
    return {
      isValid: false,
      error: 'Please enter a valid email address.'
    };
  }

  const domainLower = domain.toLowerCase();

  // Strict check for Gmail domain
  if (domainLower.startsWith('gmail.') || domainLabels[0].toLowerCase() === 'gmail') {
    if (domainLower !== 'gmail.com') {
      return {
        isValid: false,
        error: 'Please enter a valid email address.'
      };
    }
  }

  const normalized = `${localPart}@${domainLower}`;

  return {
    isValid: true,
    normalized
  };
};

export const normalizeEmail = (email) => {
  if (!email || typeof email !== 'string') return '';
  return email.trim().toLowerCase();
};

File: backend/users/validators.py
import re
from rest_framework import serializers

def validate_and_normalize_email(raw_email):
    """
    Validates and normalizes an email address.
    
    Rules enforced:
    1. Trims leading and trailing whitespace.
    2. Rejects if internal whitespace is present (e.g. 'user @gmail.com').
    3. Requires exactly one '@' character.
    4. Local-part:
       - 1 to 64 chars.
       - Cannot start or end with a dot.
       - Cannot contain consecutive dots ('..').
       - Only allowed characters: letters, numbers, and standard symbols.
    5. Domain:
       - 3 to 255 chars.
       - Cannot start or end with dot or hyphen.
       - Cannot contain consecutive dots ('..').
       - Must contain at least one dot separating domain labels and TLD.
       - Each domain label must be alphanumeric or hyphen (not starting/ending with hyphen).
    6. TLD (Top-Level Domain):
       - Must consist only of letters (2 to 24 chars, e.g. .com, .org, .in, .co.uk).
    7. Gmail-specific rule:
       - If domain is 'gmail.*' or base label is 'gmail', it MUST be strictly 'gmail.com'.
       - Disallows '.comt', '.co', '.com123', '.con', etc.
    8. Normalizes domain/email to lowercase and returns the cleaned email.
    """
    if not raw_email or not isinstance(raw_email, str):
        raise serializers.ValidationError("Email address is required.")
    
    email = raw_email.strip()
    if not email:
        raise serializers.ValidationError("Email address is required.")
        
    if re.search(r'\s', email):
        raise serializers.ValidationError("Please enter a valid email address.")
        
    if email.count('@') != 1:
        raise serializers.ValidationError("Please enter a valid email address.")
        
    local_part, domain = email.split('@')
    
    if not local_part or not domain:
        raise serializers.ValidationError("Please enter a valid email address.")
        
    # Check local-part
    if len(local_part) > 64:
        raise serializers.ValidationError("Please enter a valid email address.")
        
    if local_part.startswith('.') or local_part.endswith('.') or '..' in local_part:
        raise serializers.ValidationError("Please enter a valid email address.")
        
    local_regex = r'^[a-zA-Z0-9.!#$%&\'*+/=?^_`{|}~-]+$'
    if not re.match(local_regex, local_part):
        raise serializers.ValidationError("Please enter a valid email address.")
        
    # Check domain
    if len(domain) > 255:
        raise serializers.ValidationError("Please enter a valid email address.")
        
    if domain.startswith('.') or domain.endswith('.') or domain.startswith('-') or domain.endswith('-') or '..' in domain:
        raise serializers.ValidationError("Please enter a valid email address.")
        
    if '.' not in domain:
        raise serializers.ValidationError("Please enter a valid email address.")
        
    domain_labels = domain.split('.')
    for label in domain_labels:
        if not label or label.startswith('-') or label.endswith('-'):
            raise serializers.ValidationError("Please enter a valid email address.")
        if not re.match(r'^[a-zA-Z0-9-]+$', label):
            raise serializers.ValidationError("Please enter a valid email address.")
            
    tld = domain_labels[-1]
    # TLD must be alphabetic and between 2 and 24 characters
    if not re.match(r'^[a-zA-Z]{2,24}$', tld):
        raise serializers.ValidationError("Please enter a valid email address.")
        
    domain_lower = domain.lower()
    
    # Gmail-specific strict domain check
    if domain_lower.startswith('gmail.') or domain_labels[0].lower() == 'gmail':
        if domain_lower != 'gmail.com':
            raise serializers.ValidationError("Please enter a valid email address.")
            
    # Normalize email: lowercase
    normalized_email = f"{local_part}@{domain_lower}".lower()
    return normalized_email
